src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardNativeWrapper.cs

```
@@ -0,0 +1,292 @@
+
+        // private methods
+        private void ConfigureNativeBuffer(NativeBufferInfo buffer, PinnedBuffer pinnedBuffer, ulong size)
+        {
+            buffer.DataPointer = pinnedBuffer?.IntPtr ?? IntPtr.Zero;
+            buffer.Size = size;
+            buffer.Position = 0;
+        }
+
+        private void InitializeIfNotAlreadyInitialized()
+        {
+            if (!_operationInitialized)
+            {
+                _operationInitialized = true;
+
+                switch (_compressionMode)
+                {
+                    case CompressionMode.Compress:
+                        Zstandard64NativeMethods.ZSTD_initCStream(_zstreamPointer, _compressionLevel); // start a new compression operation
+                        break;
+                    case CompressionMode.Decompress:
+                        Zstandard64NativeMethods.ZSTD_initDStream(_zstreamPointer); // start a new decompression operation
+                        break;
+                }
+            }
+        }
+
+        // nested types
+        internal class OperationContext : IDisposable
+        {
+            private readonly PinnedBuffer _uncompressedPinnedBuffer;
+            private readonly PinnedBuffer _compressedPinnedBuffer;
+
+            public OperationContext(PinnedBuffer uncompressedPinnedBuffer, PinnedBuffer compressedPinnedBuffer)
+            {
+                _uncompressedPinnedBuffer = uncompressedPinnedBuffer; // can be null
+                _compressedPinnedBuffer = Ensure.IsNotNull(compressedPinnedBuffer, nameof(compressedPinnedBuffer));
+            }
+
+            public PinnedBuffer UncompressedPinnedBuffer => _uncompressedPinnedBuffer; // external data
+            public PinnedBuffer CompressedPinnedBuffer => _compressedPinnedBuffer; // internal data
+
+            // public methods
+            public void Dispose()
+            {
+                _uncompressedPinnedBuffer?.Dispose(); // PinnedBuffer.Dispose suppress all errors inside
```

Perhaps
`// PinnedBuffer.Dispose suppresses all errors`
?

---

done